core/src/main/scala/org/http4s/RequestInfo.scala

```diff
@@ -0,0 +1,120 @@
+  *       do not (and can not) have a [[cats.kernel.Order]] instance. If they
+  *       were included here, then we could not write a [[cats.kernel.Order]]
+  *       instance for [[RequestInfo]], limiting some of its utility, e.g. it
+  *       could not be used in a [[cats.data.NonEmptySet]].
+  */
+sealed abstract class RequestInfo extends Product with Serializable {
+  def headers: Headers
+  def httpVersion: HttpVersion
+  def method: Method
+  def uri: Uri
+
+  def withHeaders(value: Headers): RequestInfo
+  def withHttpVersion(value: HttpVersion): RequestInfo
+  def withMethod(value: Method): RequestInfo
+  def withUri(value: Uri): RequestInfo
+
+  // final //
+
+  final def mapHeaders(f: Headers => Headers): RequestInfo =
+    withHeaders(f(headers))
+
+  final def mapHttpVersion(f: HttpVersion => HttpVersion): RequestInfo =
+    withHttpVersion(f(httpVersion))
+
+  final def mapMethod(f: Method => Method): RequestInfo =
+    withMethod(f(method))
+
+  final def mapUri(f: Uri => Uri): RequestInfo =
+    withUri(f(uri))
+
+  final override def toString: String =
+    s"RequestInfo(headers = ${headers
+      .redactSensitive()}, httpVersion = ${httpVersion}, method = ${method}, uri = ${uri})"
+}
+
+object RequestInfo {
+  private[this] final case class RequestInfoImpl(
+      override final val headers: Headers,
+      override final val httpVersion: HttpVersion,
+      override final val method: Method,
+      override final val uri: Uri
+  ) extends RequestInfo {
+    override final def withHeaders(value: Headers): RequestInfo =
+      this.copy(headers = value)
+
+    override final def withHttpVersion(value: HttpVersion): RequestInfo =
+      this.copy(httpVersion = value)
+
+    override final def withMethod(value: Method): RequestInfo =
+      this.copy(method = value)
+
+    override final def withUri(value: Uri): RequestInfo =
+      this.copy(uri = value)
+  }
+
+  def apply(
+      headers: Headers,
+      httpVersion: HttpVersion,
+      method: Method,
+      uri: Uri
+  ): RequestInfo =
+    RequestInfoImpl(
+      headers,
+      httpVersion,
+      method,
+      uri
+    )
+
+  def fromRequest[F[_]](value: Request[F]): RequestInfo =
+    RequestInfoImpl(
+      value.headers,
+      value.httpVersion,
+      value.method,
+      value.uri
+    )
+
+  implicit val catsHashAndOrderForRequestInfo: Hash[RequestInfo] with Order[RequestInfo] =
+    new Hash[RequestInfo] with Order[RequestInfo] {
+      override def hash(x: RequestInfo): Int = x.hashCode
+
+      override def compare(x: RequestInfo, y: RequestInfo): Int =
+        reduceComparisons(
+          x.headers.compare(y.headers),
+          Eval.later(x.httpVersion.compare(y.httpVersion)),
+          Eval.later(x.method.compare(y.method)),
+          Eval.later(x.uri.compare(y.uri))
+        )
+    }
+
+  implicit val catsShowForRequestInfo: Show[RequestInfo] =
+    Show.fromToString
+
+  implicit val stdLibOrdering: Ordering[RequestInfo] =
```

Comment:
I think I usually implement `Ordered`, but I can't articulate why.  Do you have a reason for a preference for `Ordering`?

Reply:
Using `Ordering` allows you to write more general functions. 

This is because all types which implement `Ordered` have an `Ordering` instance, but the inverse is not true. There _is_ an implicit conversion on the companion object of `Ordered` which gives you an `Ordered` from an `Ordering`, but you can't use it as a constraint. For example, given these data types, `Foo` subtyping `Ordered[Foo]` and `Bar` implementing `Ordering[Bar]`,

```scala
  final case class Foo(value: Int) extends Ordered[Foo] {
    override final def compare(that: Foo): Int =
      value.compare(that.value)
  }

  final case class Bar(value: Int)

  object Bar {
    implicit val orderingInstance: Ordering[Bar] =
      new Ordering[Bar] {
        override def compare(x: Bar, y: Bar): Int =
          x.value.compare(y.value)
      }
  }
```

We run into this problem if we try to write a generic function on types that are orderable if we used `Ordered[A]`.

```scala
def usingOrdered[A <: Ordered[A]](x: A, y: A): Int =
  x.compare(y)

scala> val foo: Foo = Foo(1)
val foo: Foo = Foo(1)

scala> val bar: Bar = Bar(1)
val bar: Bar = Bar(1)

scala> usingOrdered(foo, foo)
usingOrdered(foo, foo)
val res0: Int = 0

scala> import Ordered.orderingToOrdered

scala> usingOrdered(bar, bar)
usingOrdered(bar, bar)
       ^
       error: inferred type arguments [Test.Bar] do not conform to method usingOrdered's type parameter bounds [A <: Ordered[A]]
                    ^
       error: type mismatch;
        found   : Test.Bar
        required: A
                         ^
       error: type mismatch;
        found   : Test.Bar
        required: A
```

The `usingOrdering(bar, bar)` fails because even though we can get an `Ordered[Bar]` just fine via the implicit conversion, `Bar` still doesn't subtype `Ordered[Bar]`. The `Ordered[Bar]` we get also has the wrong arity and semantics for use as a constraint as it is pre-applied to whatever `Bar` was used for the implicit conversion.

However, this works for both `Foo` and `Bar`,

```scala
def usingOrdering[A: Ordering](x: A, y: A): Int =
  Ordering[A].compare(x, y)

scala> usingOrdering(foo, foo)
usingOrdering(foo, foo)
val res2: Int = 0

scala> usingOrdering(bar, bar)
usingOrdering(bar, bar)
val res3: Int = 0
```

So even if we did subtype `Ordered[RequestInfo]` here, we'd probably still want to have an `Ordering[RequestInfo]`.

Reply:
I can't think offhand of an '<: Ordered[A]` constraint in the standard library, vs. several `: Ordering`s, which I think is correct.  I had a misconception that most standard library types directly implemented `Ordered` to implicitly became `Ordering`s, but they are few.  (In fact, http4s has about as many as the standard library!)

I'm good with `Ordering`, but we might want to make that consistent in 1.0.

Reply:
https://github.com/http4s/http4s/issues/5017